src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependencyModel.cs

```diff
@@ -19,6 +19,29 @@ private enum DependencyFlags : byte
             Visible  = 1 << 3
         }
 
+        protected readonly struct DependencyFlagCache
+        {
+            private readonly ProjectTreeFlags[] _lookup;
+
+            public DependencyFlagCache(ProjectTreeFlags add = default, ProjectTreeFlags remove = default)
+            {
+                ProjectTreeFlags resolved = DependencyTreeFlags.GenericResolvedDependencyFlags.Union(add).Except(remove);
+                ProjectTreeFlags unresolved = DependencyTreeFlags.GenericUnresolvedDependencyFlags.Union(add).Except(remove);
+
+                // Cannot remove implicit dependencies
+
+                _lookup = new ProjectTreeFlags[4];
+                _lookup[Index(true,  false)] = resolved;
+                _lookup[Index(true,  true )] = resolved.Except(DependencyTreeFlags.SupportsRemove);
+                _lookup[Index(false, false)] = unresolved;
+                _lookup[Index(false, true )] = unresolved.Except(DependencyTreeFlags.SupportsRemove);
+            }
+
+            public ProjectTreeFlags Get(bool isResolved, bool isImplicit) => _lookup[Index(isResolved, isImplicit)];
+
+            private static int Index(bool isResolved, bool isImplicit) => (isResolved ? 2 : 0) | (isImplicit ? 1 : 0);
```

Comment:
It might just be me, but this is very difficult to understand. I don't think some comments or maybe even an if statement or two, would hurt here. A binary literal perhaps?

Reply:
Thanks. Added some documentation.